core/src/main/java/org/infinispan/interceptors/impl/CompatibilityInterceptor.java

```diff
@@ -0,0 +1,43 @@
+package org.infinispan.interceptors.impl;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.infinispan.commands.MetadataAwareCommand;
+import org.infinispan.commands.write.ReplaceCommand;
+import org.infinispan.container.versioning.VersionGenerator;
+import org.infinispan.context.InvocationContext;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.interceptors.BasicInvocationStage;
+import org.infinispan.interceptors.DDAsyncInterceptor;
+import org.infinispan.metadata.Metadata;
+
+/**
+ * Interceptor installed when compatiblity is enabled.
+ * @author wburns
+ * @since 9.0
+ */
+public class CompatibilityInterceptor extends DDAsyncInterceptor {
+
+   protected VersionGenerator versionGenerator;
+
+   @Inject
+   protected void init(VersionGenerator versionGenerator) {
+      this.versionGenerator = versionGenerator;
+   }
+
+   @Override
+   public BasicInvocationStage visitReplaceCommand(InvocationContext ctx, ReplaceCommand command) throws Throwable {
+      addVersionIfNeeded(command);
+      return super.visitReplaceCommand(ctx, command);
+   }
+
+   protected void addVersionIfNeeded(MetadataAwareCommand cmd) {
+      Metadata metadata = cmd.getMetadata();
+      if (metadata.version() == null) {
+         Metadata newMetadata = metadata.builder()
+               .version(versionGenerator.generateNew())
+               .build();
+         cmd.setMetadata(newMetadata);
```

Comment:
I wish I knew, this is another thing where it was like this way before, heh

https://github.com/infinispan/infinispan/blob/f9945e7707ca107c9ce0bfaf8808103685300d36/core/src/main/java/org/infinispan/interceptors/compat/BaseTypeConverterInterceptor.java#L195

I can log a JIRA though so someone can look into it further.